### PR TITLE
Menu

### PR DESCRIPTION
**Pull request by jacobrienstra**
Getting there, but wanted to push so you could play with it. Hover functionality is still a little weird, as well as auto and fixed/absolute. But the positioning and offsets seem to work. Haven't tried without an activator yet.

### CODE REVIEW

**Comment by Nitwel** [comment]
Awesome 👍 
Some thoughts on it:
- Whats the diff between `openOnClick` and `disabled`
- Do we need a open and close delay?
- Do we need a dropdown relative to the position we clicked?
- Many of these props could be replaced with css vars (Maybe have a look at the badge component how its done there)
- Thats a hell lot of code for such simple feature. 😅

**Comment by rijkvanzanten** [comment]
> Do we need a open and close delay?

Yes!

> Do we need a dropdown relative to the position we clicked?

Yes! (that's what enables right-click menu's)

> Thats a hell lot of code for such simple feature.

I think you underestimate how _simple_ this is haha 😬

**Comment by rijkvanzanten** [comment]
It's looking really good already (based on Storybook). Gotta agree with @Nitwel that it's a little hard to follow along what's going on in the source code haha. Maybe it's worth splitting some of these function groups off to separate compositions within the context of v-menu.

Two quickies that will make the Storybook side of things nicer: can you make the v-list clickable so it feels more "real", and add a transition to the v-show state of the floating menu? A transition between `transform: scaleY(0);` to `scaleY(1)` should be plenty for now (or we could try re-using the `transition-expand` you added before)

File: src/components/v-menu/v-menu.vue [hunk 1]
```diff
@@ -0,0 +1,737 @@
...
+	onBeforeMount
+} from '@vue/composition-api';
+import VButton from '../v-button';
+import useWindowSize from '@/compositions/use-window-size';
+
+export default defineComponent({
+	model: {
+		prop: 'value',
+		event: 'input'
+	},
```
Comment (rijkvanzanten) [inline src/components/v-menu/v-menu.vue]: These are the defaults; can be omitted

```suggestion
```

File: src/components/v-menu/v-menu.vue [hunk 2]
```diff
@@ -0,0 +1,737 @@
...
+	components: { VButton },
```
Comment (rijkvanzanten) [inline src/components/v-menu/v-menu.vue]: Prefer relying on the globally registered v-button. No need to re-import here.

(https://github.com/directus/app-next/pull/218/files#diff-7fb45dd238eacd39fd5aa64a94dd3f2cR35)

File: src/components/v-menu/v-menu.vue [hunk 3]
```diff
@@ -0,0 +1,737 @@
...
+			default: false
+		},
+		left: {
+			type: Boolean,
+			default: false
+		},
+		right: {
+			type: Boolean,
+			default: false
+		},
```
Comment (rijkvanzanten) [inline src/components/v-menu/v-menu.vue]: Are these implemented? Didn't see to do anything in Storybook

File: src/components/v-menu/v-menu.vue [hunk 4]
```diff
@@ -0,0 +1,737 @@
...
+		positionX: {
+			type: Number,
+			default: undefined
+		},
+		positionY: {
+			type: Number,
+			default: undefined
+		},
```
Comment (rijkvanzanten) [inline src/components/v-menu/v-menu.vue]: I'm not sure if we need these props for the MVP of this. Position based on activator (+ offsetX / offsetY bools) or mouse position sounds like it covers 99% of use cases 🤔

File: src/components/v-menu/v-menu.vue [hunk 5]
```diff
@@ -0,0 +1,737 @@
...
+		closeOnClick: {
+			type: Boolean,
+			default: false
+		},
+		closeOnContentClick: {
+			type: Boolean,
+			default: false
+		},
```
Comment (rijkvanzanten) [inline src/components/v-menu/v-menu.vue]: What's the difference between these two?

File: src/components/v-menu/v-menu.vue [hunk 6]
```diff
@@ -0,0 +1,737 @@
...
+			default: undefined
+		},
+		disabled: {
+			type: Boolean,
+			default: false
+		}
+	},
+	setup(this, props, context) {
+		const isActive = ref(!!props.value);
+		const isContentActive = ref(false);
```
Comment (rijkvanzanten) [inline src/components/v-menu/v-menu.vue]: What's the difference?

File: src/components/v-menu/v-menu.vue [hunk 7]
```diff
@@ -0,0 +1,737 @@
...
+				top: 0,
+				left: 0,
+				bottom: 0,
+				right: 0,
+				width: 0,
+				height: 0,
+				offsetTop: 0,
+				scrollHeight: 0,
+				offsetLeft: 0
+			}),
```
Comment (rijkvanzanten) [inline src/components/v-menu/v-menu.vue]: Are nested objects not reactive by default? That's lame..

File: src/components/v-menu/v-menu.vue [hunk 8]
```diff
@@ -0,0 +1,737 @@
...
+			val ? callActivate() : callDeactivate();
+		});
+		watch(() => props.positionX, updateDimensions);
+		watch(() => props.positionY, updateDimensions);
+
+		watch(isContentActive, val => (hasJustFocused.value = val));
+
+		onBeforeMount(() => {
+			hasWindow.value = typeof window !== 'undefined';
+		});
```
Comment (rijkvanzanten) [inline src/components/v-menu/v-menu.vue]: It's safe to assume this is used in a browser context. We don't have a server so no SSR either.

File: src/components/v-menu/v-menu.vue [hunk 9]
```diff
@@ -0,0 +1,737 @@
...
+		function callDeactivate() {
+			isContentActive.value = false;
+		}
+
+		/**
+		 *
+		 * Calc Position
+		 *
+		 **/
+		const calculatedLeft = computed((): string => {
```
Comment (rijkvanzanten) [inline src/components/v-menu/v-menu.vue]: You can typecast computed's as follows:

```suggestion
		const calculatedLeft = computed<string>(() => {
```

File: src/components/v-menu/v-menu.vue [hunk 10]
```diff
@@ -0,0 +1,737 @@
...
+			return dimensions.activator.left - defaultOffset.value * 2;
+		}
+
+		const calculatedTop = computed((): string => {
+			const top = !props.auto ? calcTop() : `${calcYOverflow(calculatedTopAuto.value)}px`;
+			return top || '0';
+		});
+		const computedTop = computed((): number => {
+			const a = dimensions.activator;
+			const c = dimensions.content;
```
Comment (rijkvanzanten) [inline src/components/v-menu/v-menu.vue]: Prefer explicit names for vars (you can't find this by CMD+F `a`)

```suggestion
			const activator = dimensions.activator;
			const content = dimensions.content;
```

Comment (rijkvanzanten) [reply src/components/v-menu/v-menu.vue]: (suggestion just as an example, needs to be changed elsewhere in this function too obviously)

File: src/components/v-menu/v-menu.vue [hunk 11]
```diff
@@ -0,0 +1,737 @@
...
+			contentListeners: genContentListeners
+		};
+	}
+});
+</script>
+
+<style lang="scss" scoped>
+.v-menu {
+	width: min-content;
+	max-width: min-content;
```
Comment (rijkvanzanten) [inline src/components/v-menu/v-menu.vue]: This might pose a problem for certain pieces of content 🤔 

I'd like to avoid unnecessary line breaks in things like list items if possible

Comment (rijkvanzanten) [reply src/components/v-menu/v-menu.vue]: Maybe we should make the "not-attached-to-activator" version a fixed width

File: src/directives/click-outside/click-outside.ts [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+function closeConditional() {
+	return false;
+}
+
+function directive(e: PointerEvent, el: HTMLElement, binding: ClickOutsideDirective): void {
+	// Args may not always be supplied
+	binding.args = binding.value?.args || {};
+
+	// If no closeConditional was supplied assign a default
+	const isActive = binding.args.closeConditional || closeConditional;
```
Comment (rijkvanzanten) [inline src/directives/click-outside/click-outside.ts]:
```suggestion
	const isActive = binding.args.closeConditional || false;
```

I don't think this needs a separate function

File: src/directives/click-outside/click-outside.ts [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+	// so we should avoid calling them when we're not active.
+	// Explicitly check for false to allow fallback compatibility
+	// with non-toggleable components
+	if (!e || isActive(e) === false) return;
+
+	// If click was triggered programmaticaly (domEl.click()) then
+	// it shouldn't be treated as click-outside
+	// Chrome/Firefox support isTrusted property
+	// IE/Edge support pointerType property (empty if not triggered
+	// by pointing device)
```
Comment (rijkvanzanten) [inline src/directives/click-outside/click-outside.ts]: 👏  so 👏  many 👏  edge 👏  cases 👏

File: src/directives/click-outside/click-outside.ts [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+	if (('isTrusted' in e && !e.isTrusted) || ('pointerType' in e && !e.pointerType)) return;
+
+	// Check if additional elements were passed to be included in check
+	// (click must be outside all included elements, if any)
+	const elements = (binding.args.include || (() => []))();
+	// Add the root element for the component this directive was defined on
+	elements.push(el);
```
Comment (rijkvanzanten) [inline src/directives/click-outside/click-outside.ts]: Does this allow clicking outside of the element on _specific_ other elements? Very powerful, but also a bit out of scope maybe?

File: src/directives/click-outside/click-outside.ts [hunk 15]
```diff
@@ -0,0 +1,80 @@
...
+		// or body, this is the entire purpose of the v-app
+		// component and [data-app]
+		const app = document.querySelector('[data-app]') || document.body; // This is only for unit tests
+		app.addEventListener('click', onClick, true);
+	},
+
+	unbind(el: HTMLElement, binding: ClickOutsideDirective) {
+		const onClick = (e: Event) => directive(e as PointerEvent, el, binding);
+		const app = document.querySelector('[data-app]') || document.body; // This is only for unit tests
+		app && app.removeEventListener('click', onClick, true);
```
Comment (rijkvanzanten) [inline src/directives/click-outside/click-outside.ts]: You haven't written the unit tests, so this is probably not even needed in our case

File: src/directives/click-outside/click-outside.ts [hunk 16]
```diff
@@ -0,0 +1,80 @@
...
+	// available, iOS does not support
+	// clicks on body
+	inserted(el: HTMLElement, binding: ClickOutsideDirective) {
+		const onClick = (e: Event) => directive(e as PointerEvent, el, binding);
+		// iOS does not recognize click events on document
+		// or body, this is the entire purpose of the v-app
+		// component and [data-app]
+		const app = document.querySelector('[data-app]') || document.body; // This is only for unit tests
+		app.addEventListener('click', onClick, true);
+	},
```
Comment (rijkvanzanten) [inline src/directives/click-outside/click-outside.ts]: Our main app component doesn't have the `data-app` attribute (yet)

File: src/components/v-menu/v-menu.vue (hunk 3 above)
Comment (jacobrienstra) [reply src/components/v-menu/v-menu.vue]: Yeah!  They're there!
Screen Shot 2020-03-24 at 4 33 09 PM

File: src/components/v-menu/v-menu.vue (hunk 4 above)
Comment (jacobrienstra) [reply src/components/v-menu/v-menu.vue]: That's what I thought to, but if you want to use mouse position without activator, you need to be able to capture it yourself and pass it to the menu. Plus, it's already there and it works so I figured ¯\_(ツ)_/¯

File: src/components/v-menu/v-menu.vue (hunk 6 above)
Comment (jacobrienstra) [reply src/components/v-menu/v-menu.vue]: I believe it has to do with hovering or perhaps even focusing with keys. Will investigate.

File: src/components/v-menu/v-menu.vue (hunk 7 above)
Comment (jacobrienstra) [reply src/components/v-menu/v-menu.vue]: It....I mean....are they? By default they unpack to the values when using dot notation, but it seemed like they weren't being reactive. I could be wrong though...

File: src/components/v-menu/v-menu.vue (hunk 11 above)
Comment (jacobrienstra) [reply src/components/v-menu/v-menu.vue]: i dont actually think we need this, it was a temporary measure while i was getting the dimension calculations worked out

File: src/directives/click-outside/click-outside.ts (hunk 13 above)
Comment (jacobrienstra) [reply src/directives/click-outside/click-outside.ts]: At least! Vuetify! Caught them before we did!

File: src/directives/click-outside/click-outside.ts (hunk 16 above)
Comment (jacobrienstra) [reply src/directives/click-outside/click-outside.ts]: What's the root app element?

File: src/directives/click-outside/click-outside.ts (hunk 14 above)
Comment (jacobrienstra) [reply src/directives/click-outside/click-outside.ts]: No, I don't think so. It's just a way of passing in all the elements it should NOT listen for clicks on, i.e., all the ones involved in the element. In menu's case, this means both the activator and the content.

File: src/components/v-menu/v-menu.vue (hunk 5 above)
Comment (jacobrienstra) [reply src/components/v-menu/v-menu.vue]: Close on click turns on the outside-click thing.

File: src/components/v-menu/v-menu.vue (hunk 7 above)
Comment (rijkvanzanten) [reply src/components/v-menu/v-menu.vue]: No idea!

File: src/directives/click-outside/click-outside.ts (hunk 16 above)
Comment (rijkvanzanten) [reply src/directives/click-outside/click-outside.ts]: You can use either body directly or `div#app`

File: src/components/v-menu/v-menu.vue (hunk 6 above)
Comment (jacobrienstra) [reply src/components/v-menu/v-menu.vue]: they are! isActive directly mirrors the value, isContentActive is a more gated var, i.e. disabled, click outside, focus stuff affects it.

**Comment by rijkvanzanten** [comment]
Well, that turned out to be quite the challenge, but we made it through @jacobrienstra . Go team! ⭐️ 💪
